### DISCUSSION TITLE
RuntimeError: cannot register a hook on a tensor that doesn't require gradient

### DISCUSSION THREAD

**Discussion by RuixiangZhao** [frgfm/torch-cam#146] [Q&A] [answered]
When I tried to use torchcam on my fine-tuned resnet50, it threw an error:
> WARNING:root:no value was provided for `target_layer`, thus set to 'layer4'.
Traceback (most recent call last):
  File "grad-cam.py", line 33, in <module>
    out = my_model(input_tensor.unsqueeze(0))
  File "/home/zrx/miniconda3/envs/py38/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1120, in _call_impl
    result = forward_call(*input, **kwargs)
  File "/home/zrx/stomatology/baseline/models/resnet.py", line 205, in forward
    x = self.layer4(x)
  File "/home/zrx/miniconda3/envs/py38/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1123, in _call_impl
    hook_result = hook(self, input, result)
  File "/home/zrx/torch-cam/torchcam/methods/gradient.py", line 50, in _hook_g
    self.hook_handles.append(output.register_hook(partial(self._store_grad, idx=idx)))
  File "/home/zrx/miniconda3/envs/py38/lib/python3.8/site-packages/torch/_tensor.py", line 341, in register_hook
    raise RuntimeError("cannot register a hook on a tensor that "
RuntimeError: cannot register a hook on a tensor that doesn't require gradient

The main codes are as follows：
```Python
import torch
from torchvision import transforms
from torchcam.methods import SmoothGradCAMpp
from PIL import Image

my_model = torch.load(model_path, map_location=device).eval()
cam_extractor = SmoothGradCAMpp(my_model)
img = Image.open(image_path)
test_transforms = transforms.Compose(
        [transforms.Resize(224),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])])
input_tensor = test_transforms(img).to(device)
out = my_model(input_tensor.unsqueeze(0))
...
```
And my environment verion is:
> torch==1.10.0
torchvision==0.11.1
torchcam==0.3.2.dev0

**Comment by frgfm** [comment, marked as answer]
Hi @RuixiangZhao,

Thanks for bringing up that up :pray:  
First, I'd like to point a few things out : 
- duplicates do not benefit developers, there is already a discussion #132 on this topic, it would have been best to put this information over here :pray: 
- as mentioned in the other discussion, I clearly made issue templates specifically for this purpose. Discussions are for open questions / suggestions / etc, bug reports are more relevant as GH issues. I took time to make the issue template so that you provide all the information necessary for me to help you debug this :sweat_smile:  (this includes your full relevant environment setup, and a fully executable script)

In this case:
- torch & torchcam versions are relevant, but there is a lot more (again instructions to get this information easily in one line are given in the issue template)
- your snippet cannot be executed as-is (model_path, image_path, device definition)
- I took the liberty of adapting it to something fully runnable (cf. below) and I'm not able to reproduce your error:
```python
import torch
from torchvision import transforms
from torchvision.models import resnet50
from torchcam.methods import SmoothGradCAMpp
from PIL import Image

my_model = resnet50().eval()
cam_extractor = SmoothGradCAMpp(my_model)
img = Image.fromarray((255 * torch.rand((224, 224, 3))).round().to(dtype=torch.uint8).numpy())
test_transforms = transforms.Compose(
        [transforms.Resize(224),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])])
input_tensor = test_transforms(img)
out = my_model(input_tensor.unsqueeze(0))
```
and I'm unable to reproduce it so far. 
Don't take it the wrong way, I'm happy to help :) As much as possible, I'd love to do this efficiently, so if you mind, I'd suggest to:
- close this discussion (duplicate)
- open a Github issue (just need to follow the "bug-report" template) about this
- mention the issue in the other discussion 

Would that be OK with you? :)

**Reply by RuixiangZhao** [reply]
https://github.com/frgfm/torch-cam/discussions/132#discussioncomment-2240733
